### PR TITLE
Trade UI update

### PR DESCRIPTION
**Pull request by Pink-Chink**
## About The Pull Request

New station selection screen.
image

Trade screen toolbar with cart viewing, saved carts, and ordering. Button for fulfilling all offers at the top right.
image

Carts can handle multiple stations. Ordering, saving, and loading buttons.
image

Cart menu.
image

Order requests and queue.
image

## Why It's Good For The Game

Better UI.

## Changelog
:cl:
add: Added selection tree, viewable carts, cart save slots, an ordering system, and a function for fulfilling all offers to the trade program
add: Added trade logging app
tweak: Trade program supports carts with goods from multiple stations
tweak: Trade orders can be done via the catalog/ordering app on the PDA (check downloadable apps)
balance: increased non-offer favor multiplier to 25%
/:cl:

### CODE REVIEW

**Comment by ValoTheValo** [comment]
Why remove the ability to purchase from the front desk console. thats just annoying to deal with

**Comment by MLGTASTICa** [comment]
> Why remove the ability to purchase from the front desk console. thats just annoying to deal with

I fully agree with what Valo said , this just makes their usage super annoying

**Comment by garrulousTeamotei** [comment]
The Station Research Tree looks wonderful, buying from multiple stations is superb, and holding invoices is a nice touch, but why the restriction on the trade beacon?
If the trade program came with the Guild account by default, I could understand. All it seems it's preventing is someone else from buying things with their own money, things they couldn't pick up without guild access anyways.

**Comment by Pink-Chink** [comment]
Currently, the beacons can be accessed from any trade program regardless of location. The area restriction is to prevent weirdness with multiple beacons and abuse/theft. On Sojourn, they had multiple beacons set up and relied on honor rules to prevent theft. This just lets that be mechanically enforced. As far as abuse/theft goes, there is a disk next to the regular trading console down by the beacons. Someone with a laptop could break in, steal the disk, and hijack offers with their own account from some maint fort. It sounds on-theme for Eris, but it doesn't pan out in practice. The only feedback you get is from the logs and the beacon sparks, but I doubt anyone will be looking at those closely. If there were some meaningful counterplay or a better indicator, I wouldn't really give a shit.

**Comment by MLGTASTICa** [comment]
> 

You could let people configure trade beacons to be only accesible by people with X acces codes.

**Comment by Pink-Chink** [comment]
Doable, but that seems a bit unnecessary. The catalog version of the trade program still lets the person at the desk build and queue orders using the Guild account.

**Comment by ValoTheValo** [comment]
>On Sojourn, they had multiple beacons set up and relied on honor rules to prevent theft. This just lets that be mechanically enforced

This isnt sojourn. Nobody ever does stuff like that. Most ive seen is the captain occasionally hooking up thier own account. And if someone does break in past the large number of armed guildies and orders stuff for themselves and doesnt die. Good for them.
All i see that change accomplishing is making guild less fun to play and more annoying to work with as nobody will be manning the desk

**Comment by thedarkcrusader** [comment]
> > On Sojourn, they had multiple beacons set up and relied on honor rules to prevent theft. This just lets that be mechanically enforced
> 
> This isnt sojourn. Nobody ever does stuff like that. Most ive seen is the captain occasionally hooking up thier own account. And if someone does break in past the large number of armed guildies and orders stuff for themselves and doesnt die. Good for them. All i see that change accomplishing is making guild less fun to play and more annoying to work with as nobody will be manning the desk

I don't have really anything meaningful to bring to the discussion but I do have one thing. I don't actually understand how it makes guild unf7n to play when really all you got to do is go down 1 flight of stairs and confirm the order on the computer  doesn't really change much OF IMO

**Comment by ValoTheValo** [comment]
> > > On Sojourn, they had multiple beacons set up and relied on honor rules to prevent theft. This just lets that be mechanically enforced
> > 
> > 
> > This isnt sojourn. Nobody ever does stuff like that. Most ive seen is the captain occasionally hooking up thier own account. And if someone does break in past the large number of armed guildies and orders stuff for themselves and doesnt die. Good for them. All i see that change accomplishing is making guild less fun to play and more annoying to work with as nobody will be manning the desk
> 
> I don't have really anything meaningful to bring to the discussion but I do have one thing. I don't actually understand how it makes guild unf7n to play when really all you got to do is go down 1 flight of stairs and confirm the order on the computer doesn't really change much really IMO

you need to do that every time you want to order something and people on eris move at the speed of snails most of the time. its just needless tedious for no good reason

**Comment by garrulousTeamotei** [comment]
What really needs to he done is to put the beacons on the 3rd z level instead of the first. One of the main reasons for the current mapping is the intended elevator feature. The elevator that was never added and generally forgot about. If only that was fixes during the NT/Guild remapping challenge.

**Comment by Voldirs** [comment]
> Currently, the beacons can be accessed from any trade program regardless of location. The area restriction is to prevent weirdness with multiple beacons and abuse/theft. On Sojourn, they had multiple beacons set up and relied on honor rules to prevent theft. This just lets that be mechanically enforced. As far as abuse/theft goes, there is a disk next to the regular trading console down by the beacons. Someone with a laptop could break in, steal the disk, and hijack offers with their own account from some maint fort. It sounds on-theme for Eris, but it doesn't pan out in practice. The only feedback you get is from the logs and the beacon sparks, but I doubt anyone will be looking at those closely. If there were some meaningful counterplay or a better indicator, I wouldn't really give a shit.

If that's concern, those beacons could have passwords set when being constructed and those passwords could be set in guildie's memories so they can use their beacons more conveniently and all default consoles would have beacons set by default. I don't know. It sounds not so great too, but on Eris it's not a concern for now, but your decision to fix the problem that's not even here makes it MUCH less convenient to use.

**Comment by Pink-Chink** [comment]
As the commit says, I've reverted the console back. Beacon area restrictions won't apply to the master program (the one you guys are familiar with).

**Comment by Pink-Chink** [comment]
Also, I'm adding the ordering app to the list of default PDA apps. I figured it would be better than having to remind people to download it every round.

**Review by TheShown911 — APPROVED** [review]
Wow, a big feature PR that makes the game more interesting. Good job. I can only hope other PRs from this week are like this...

**Comment by Voldirs** [comment]
> Trade screen toolbar with cart viewing, saved carts, and ordering. Button for fulfilling all offers at the top right

Damn, I forgot one thing when I posted that suggestion that needs to be tested, what happens when there is the item that fills 2 contracts at once, such like artistic toolmod fulfilling both toolmod and artistic toolmod contract

**Comment by Pink-Chink** [comment]
It depends on the order of offer types when the list gets built, but I should probably set it up so that specific types get filled first no matter what. I'll make a fix for that when I'm done with organs.

**Comment by Humonitarian** [comment]
Can I get a short rundown on how ordering works?

**Comment by Pink-Chink** [comment]
Player makes a cart on the ordering app
Player submits order request with app
Guild reviews/approves
Trade system charges Guild for order
Guild charges Player for order plus handling fee (both transactions happen automatically when Guild hits “purchase”)
Crate with order is created at the beacon

ID locked crates would be a good idea. Currently, the only way to know is to download the trade logs app and check.

**Comment by Humonitarian** [comment]
^ yeah, ID lock, but hackable with hacking device / emag

File: nano/templates/trade_goods.tmpl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	{{else}}
+		{{:helper.link(data.total, "comment", {"PRG_build_order" : 1}, data.total ? null : 'disabled')}}
+	{{/if}}
+	{{:helper.link('Reset', 'trash', {'PRG_cart_reset' : 1})}}
+	{{:helper.link("View Cart", 'search', {'PRG_trade_screen' : "cart"})}}
+{{else}}
+	<div class='nanoMap' style='padding: 100px 200px;'>
+		Select a trade station.
+	</div>
+{{/if}}
```
Comment (hyperioo) [inline nano/templates/trade_goods.tmpl:45]:
```suggestion
{{/if}}

```

File: nano/templates/trade_offers.tmpl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+					{{if data.prg_type != "ordering"}}
+						<td style="width: 0">{{:helper.link("", 'check', {'PRG_offer_fulfill' : value.index, 'PRG_offer_fulfill_path' : value.path}, (value.available >= value.amount && value.amount > 0) ? null : 'disabled')}}
+					{{/if}}
+				{{/if}}
+			{{/for}}
+		</table>
+	{{else}}
+		No offers found.
+	{{/if}}
+</div>
```
Comment (hyperioo) [inline nano/templates/trade_offers.tmpl:33]:
```suggestion
</div>

```

File: nano/templates/trade_orders.tmpl [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+			</div>
+		<td>
+			<div style="padding: 0 0 0 40%">
+				{{:helper.link("", 'arrowthick-1-e', {'PRG_page_next_10' : data.page_max}, data.order_page == data.page_max ? 'disabled' : null)}}
+			</div>
+		<td>
+			<div style="padding: 0 0 0 40%">
+				{{:helper.link("", 'arrowthickstop-1-e', {'PRG_page_last' : data.page_max}, data.order_page == data.page_max ? 'disabled' : null)}}
+			</div>
+</table>
```
Comment (hyperioo) [inline nano/templates/trade_orders.tmpl:137]:
```suggestion
</table>

```

File: nano/templates/trade_saved.tmpl [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		<td>
+			<div style="padding: 0 0 0 35%">
+				{{:helper.link("", 'arrowthick-1-e', {'PRG_page_next_10' : "saved"}, data.cart_page == data.page_max ? 'disabled' : null)}}
+			</div>
+		<td>
+			<div style="padding: 0 0 0 35%">
+				{{:helper.link("", 'arrowthickstop-1-e', {'PRG_page_last' : "saved"}, data.cart_page == data.page_max ? 'disabled' : null)}}
+			</div>
+</table>
+</div>
```
Comment (hyperioo) [inline nano/templates/trade_saved.tmpl:45]:
```suggestion
</div>

```

**Review by hyperioo — COMMENTED** [review]
Great PR, good job!

**Pull request merged**